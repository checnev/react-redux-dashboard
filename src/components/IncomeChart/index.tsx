import React from 'react';
import {
  AreaChart,
  Tooltip,
  XAxis,
  YAxis,
  CartesianGrid,
  Area,
  ResponsiveContainer,
} from 'recharts';
import { Typography } from '@material-ui/core';
import { makeStyles, createStyles, Theme, useTheme } from '@material-ui/core/styles';
import { Skeleton } from '@material-ui/lab';

import type { DailyReport } from 'selectors/stats';

const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {
    overflow: 'hidden',
  },
  title: {
    display: 'flex',
    flexGrow: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: theme.spacing(1),
  },
  titleDate: {
    fontSize: '.75rem',
  }
}));

interface IncomeChartProps {
  reports: DailyReport[];
}

const IncomeChart: React.FC<IncomeChartProps> = ({ 
  reports,
} : IncomeChartProps) => {
  const classes = useStyles();
  const theme = useTheme();
  return (
    <div className={classes.root}>
      <div className={classes.title}>
        <Typography
          color="primary"
          variant="h6"
          component="h2"
        >
          Доход
        </Typography>
        <Typography
          align="right"
          color="textSecondary"
          className={classes.titleDate}
        >
          {!reports.length && <Skeleton variant="text" width={50} />}
          {reports.length > 0 && (
            <span>
              {reports[0].date} - {reports[reports.length - 1].date}
            </span>
          )}
        </Typography>
      </div>
      {!reports.length && <Skeleton variant="rect" width="100%" height={200} />}
      {reports.length > 0 && (
        <ResponsiveContainer width="100%" height={200}>
          <AreaChart 
            data={reports as object[]}
            margin={{ top: 10, right: 30, left: 0, bottom: 0 }}
          >
            <XAxis dataKey="date" />
            <YAxis
              label={{ 
                value: 'Доход ₽',
                angle: -90,
                position: 'insideLeft',
                fill: theme.palette.text.secondary,
              }}
            />
            <CartesianGrid strokeDasharray="3 3" />
            <Tooltip labelStyle={{color: 'rgba(0, 0, 0, .87)'}}/>
            <Area type="monotone" dataKey="income" name="Доход" fill="none" stroke="#8884d8" strokeWidth="2" />
            <Area type="monotone" dataKey="countOfLeases" name="Кол-во Аренд" fill="none" stroke="#03f442" strokeWidth="0" />
            <Area type="monotone" dataKey="incidents" fill="none" name="Происшествий" stroke="#f44336" strokeWidth="0" />
          </AreaChart>
        </ResponsiveContainer>
      )}
    </div>
  );
};

export default IncomeChart;
