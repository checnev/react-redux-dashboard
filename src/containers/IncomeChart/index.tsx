import { useSelector } from 'react-redux';
import { dailyReportsSelector } from 'selectors/stats';
import IncomeChart from 'components/IncomeChart';

const IncomeChartContainer= () => {
  const incomeReports = useSelector(dailyReportsSelector);
  
  return <IncomeChart reports={incomeReports} />;
};

export default IncomeChartContainer;
