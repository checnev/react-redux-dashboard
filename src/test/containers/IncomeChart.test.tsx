import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import IncomeChart from 'containers/IncomeChart';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import state from 'test/data/initialState';

const mockStore = configureMockStore([thunk]);

jest.mock('recharts', () => ({
  AreaChart: () => <div data-testid="AreaChart"></div>,
  Area: () => <div data-testid="area"></div>,
  Tooltip: () => <div data-testid="Tooltip"></div>,
  XAxis: () => <div data-testid="XAxis"></div>,
  YAxis: () => <div data-testid="YAxis"></div>,
  CartesianGrid: () => <div data-testid="CartesianGrid"></div>,
  ResponsiveContainer: () => <div data-testid="ResponsiveContainer"></div>,
}));

let container: any = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('IncomeChartContainer render', () => {
  const store = mockStore(state);
  act(() => {
    render(
      <Provider store={store}>
        <IncomeChart />
      </Provider>,
      container
    );
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
