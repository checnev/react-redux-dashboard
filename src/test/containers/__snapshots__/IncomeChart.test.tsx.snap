// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeChartContainer render 1`] = `
"<div class=\\"makeStyles-root-1\\">
  <div class=\\"makeStyles-title-2\\">
    <h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary\\">Доход</h2>
    <p class=\\"MuiTypography-root makeStyles-titleDate-3 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-alignRight\\"><span>10.12.2020 - 16.12.2020</span></p>
  </div>
  <div data-testid=\\"ResponsiveContainer\\"></div>
</div>"
`;
